tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/op/core/ShapeOps.java

```diff
@@ -337,4 +336,252 @@
+     *
+     * @param scope current scope
+     * @param shape the TensorFlow shape
+     * @param n     the number of leading dimensions to get, must be <= than the shape's numDimensions()
+     * @param dType the shape datatype.
+     * @param <U>   the shape datatype.
+     * @return an n-dimensional Shape with the first n dimensions matching the first n dimensions
+     * of this Shape
+     */
+    @Endpoint(name = "takeLast")
+    public static <U extends TNumber> Operand<U> takeLast(
+            Scope scope, Shape<U> shape, Operand<U> n, DataType<U> dType) {
+
+        Size<U> rank = Size.create(scope, shape, dType);
+        Sub<U> start = Sub.create(scope, rank, n);
+        Operand<U> dims = Slice.create(scope, shape,
+                ExpandDims.create(scope, start, Constant.scalarOf(scope, -1)),
+                ExpandDims.create(scope,
+                        Cast.create(scope, Constant.scalarOf(scope, -1), dType),
+                        Constant.scalarOf(scope, -1)));
+        return dims;
+    }
+
+    /**
+     * Returns a new Shape, with a new first dimension inserted.
+     *
+     * @param scope          current scope
+     * @param shape          the TensorFlow shape
+     * @param firstDimension the dimension to prepend
+     * @return a new shape with the given dimension(s) first, followed by the Shape's dimensions.
+     */
+    @Endpoint(name = "prepend")
+    public static Operand<TInt32> prepend(
+            Scope scope, Shape<TInt32> shape, int firstDimension) {
+        Operand<TInt32> dim = Constant.arrayOf(scope, new int[]{firstDimension});
+        Concat<TInt32> result = Concat.create(scope,  Arrays.asList(dim, shape),
+                Constant.scalarOf(scope, 0));
+
+        return result;
+    }
+
+    /**
+     * Returns a new Shape, with a new first dimension inserted.
+     *
+     * @param scope          current scope
+     * @param shape          the TensorFlow shape
+     * @param firstDimension the dimension to prepend
+     * @return a new shape with the given dimension(s) first, followed by this Shape's dimensions.
+     */
+    @Endpoint(name = "prepend")
+    public static Operand<TInt64> prepend(
+            Scope scope, Shape<TInt64> shape, long firstDimension) {
+        Operand<TInt64> dim = Constant.arrayOf(scope, new long[]{firstDimension});
+        Concat<TInt64> result = Concat.create(scope,  Arrays.asList(dim, shape),
+                Constant.scalarOf(scope, 0));
+
+        return result;
+    }
+
+
+
+    /**
+     * Returns a new Shape, with another Shape's dimensions prepended.
+     *
+     * @param scope          current scope
+     * @param shape          the TensorFlow shape
+     * @param otherShape     the other shape to prepend
+     * @return a new shape with another Shape's dimensions prepended.
+     */
+    @Endpoint(name = "prepend")
+    public static <T extends TNumber> Operand<T> prepend(
+            Scope scope, Operand<T> shape, Operand<T> otherShape) {
```

Comment:
should the `Operand` here be `Shape` instead? I think we need to use strong typing here otherwise a user might easily think that `tf.append` can be used for any kind of tensors.

Reply:
I assume you are referring to `org.tensorflow.op.core.Shape`.
These new operations manipulate  "dynamic shapes" as detailed in the python code, as opposed to "static shapes" represented by `org.tensorflow.tools.Shape`. The primary use of this change is in creating and manipulating shapes within the TF operations, outside of Java code. Dynamic shape manipulations usually results because the static Shape has unknown dimensions or the shape needs to be manipulated within the graph to be compatible with another Graph Operation.

`org.tensorflow.op.core.Shape` is a specific kind of operand that contains dimensions, 
but any Operand that translates into a set of numbers (dimensions) works.
For example,` tf.constant(new int[] { 1,2,3}), tf.slice(....), tf.concat(), tf.stack(), tf.shape(operand) or tf.constant(org.tensorflow.tools.Shape)` all work equally. Also, the output of these all work with tf.reshape(),
defined as `reshape(Operand<T> tensor, Operand<U> shape)`.

As far as I can tell, there is no way to create a `org.tensorflow.op.core.Shape` object from Operands containing dimensions. `tf.shape()` only takes the shape associated with  the Operand, itself. Perhaps we need an operand that creates a `org.tensorflow.op.core.Shape` object from dimensions,  but using an Operand of dimensions seems to suffice, at least at the level that I need to  use these functions.

This PR creates three signatures for append, all for manipulating the dimensions of a "shape".

> Operand<T> append(Operand<T> shape, Operand<T> otherShape)
> Operand<TInt64> append(Shape<TInt64> shape,  long lastDimension)
> Operand<TInt32> append(Shape<TInt32> shape,  int lastDimension)

The Shape<TIntxx> versions os these methods require a Shape object but they are really a convenience in passing
the lastDimension as a Java int/long. Calling `append(shapeObj, tf.constant(new int/long[]i) )`
produces the same result, but is not as straight forward.

Reply:
Ok I understand your point, it is in case the given `shape` is any kind of tensor and not necessarily the result of the `Shape` op, make sense. 

Maybe we should update the documentation then because it stated that all the endpoint returns "a Shape" while it simply returns any integer rank-1 tensor that could be used as a shape or not (it looks like nothing would prevent a user to use this endpoint for any other reason).

If we want to make it clear that it should be used mainly for shapes, maybe we should add a prefix to all this endpoints so that they are accessible via `tf.shape.prepend` instead of just `tf.prepend`? That would be possible by simply adding a group to the operator annotation, i.e. `@Operator(group="shape")`

Reply:
I added the `@Operator(group="shape")`, I was thinking of doing this anyway as it was growing bigger that I had originally thought and now it makes it clear what it is used for.
I changed the JavaDoc and some other parameters to  clarify that Operand<> of dimensions are used rather than Shape objects. I reformatted with google-java-format formatter on both ShapeOps and ShapeOpsTest. Also cleaned up ShapeOps based on InteliJ Code Analysis. Changes are pushed.

Reply:
Thanks @JimClarke5 , looks good. The only thing though is that I'm pretty sure that when you ran `mvn install` on your branch, a few generated classes ended up being modified or added under `src/gen/`. These need to be committed as part of this PR too. 

Also, probably that one of the file created is called `ShapeOps` as well and found under `src/gen/annotations/org/tensorflow/op`. It won't conflict with yours but it might be confusing to have two classes with the same name in the same jar. So I suggest then that you rename your operator class just `Shapes` (and `ShapesTest`). 

Note: at some point we will probably also need to rename the `core` package found in `src/main/java/org/tensorflow/op` as it conflicts with the same package found under `src/gen/java` but we can deal with this another time.

Reply:
Ok, I have renamed `ShapeOps` ==> `Shapes` and `ShapeOpsTest` ==> `ShapesTest` under main.
`ShapeOps` under gen is generated from "Shapes.java", so I assumed (wrongly) that it wasn't tracked in Git,
so I have added it and Ops.java back into my local repository.

FYI, from my experience with Java modules, post Java 8, is that it gets very finicky about duplicate package names, especially across separate jars. 

Having duplicate class names under different packages is possible, but gets nasty when using the  identically named classes within  the same code of another class, as I am finding out with the two Shape classes. Many times in the Keras code, it tries to first resolve the shape statically using `org.tensorflow.tools.Shape`, but if it can't, it then resolves the shape dynamically using `org.tensorflow.op.core.Shape`. The resolution is to fully qualify the class with the package name when using both identically named classes.

Reply:
> gets nasty when using the identically named classes within the same code of another class, as I am finding out with the two Shape classes.

I totally agree with you, I have the same surprise every time I'm working in the `org.tensorflow.op.core` package and I realize that my `Shape` instances are not of the expected type. I think it makes more sense that the static version keeps the `Shape` name and that we should override the name for the generated one (maybe `TensorShape`?). 

Let's do this another time though, this PR looks good to be merged now, thanks!